Fetching opds custom list

EdwinGuzman opened the pull request:
I think this could use a bit more refactoring (and tests) but wanted to get this out here to break out the work. It may be easier to do this and work on the front-end at the same time for rendering this opds custom list feed, and then another PR for saving and deleting entries in the custom list. That way I won't hold up Adriana, too.

aslagle commented on api/admin/controller.py:
```diff
@@ -1600,26 +1604,35 @@ def custom_list(self, list_id):
+                library_short_name=library.short_name,
+            )
+
+            feed = AcquisitionFeed(self._db, list.name, url, page_of_works, annotator)
```
You can use `AcquisitionFeed.page` and pass in the pagination so you don't have to add the links yourself.

EdwinGuzman replied:
I tried `AcquisitionFeed.page` initially but that has the issue that it queries against the materializedview so nothing gets returned to the feed.

aslagle commented on api/opds.py:
```diff
@@ -768,6 +769,19 @@ def annotate_feed(self, feed, lane):
 
+        if lane and lane.uses_customlists and (hasattr(lane, "customlist_ids") and lane.customlist_ids):
+            crawlable_url = self.url_for(
+                "crawlable_list_feed", list_name="custom list",
```
I don't get what this is for. And why is the list name "custom list"?

EdwinGuzman replied:
WorkList now has a `customlist_ids` property but a Lane does not. So checking to see if it's there in order to add the crawlable url. "custom list" is the name for now because I haven't been able to get the list name from just the array of custom list ids.

aslagle replied:
There's a `customlist_ids` property but no `customlists` property? This is almost identical to the previous code except it doesn't restrict you to lanes with one list, but this probably should too. Right now you're just making a link that will give you a 404 since it's not a real list name.

EdwinGuzman replied:
There could be either `customlist_ids` or `customlists`.
Right, I would have to do more to get the list from the id to get the name.

aslagle commented:
Another approach would be to modify the crawlable list feed controller method so that it did this stuff if the list didn't have a lane.

EdwinGuzman commented:
The `crawlable_list_feed` also works against the materialized view. That's the controller method you think can be modified?

aslagle commented:
We used to have some code in opds.py to make it work with either a MaterializedWork or a regular Work, and I think we got rid of it because we didn't need it anymore.

leonardr commented:
We had code in `Lane.works` that would generate the "find the books" query based on either Work or MaterializedWork. That code was called by `AcquisitionFeed.page` but you could also get Works some other way and just dump them into an `AcquisitionFeed`.

Now `Lane.works` always goes against MaterializedWork, and now we're finding out that this means you can't use `Lane.works` in admin interface situations. But I don't think we'd want to use it anyway, because `Lane.works` also grabs a `CachedFeed` if possible, and in an admin interface you always want to see live results. "Book's list membership isn't reflected in the materialized view" is just a more annoying version of "Book isn't in the cached feed".

Right now there aren't many parts of the application where we need to build an OPDS feed of books -- you're either searching for books (in which case you use the existing search controller) or you're looking at the books on a custom list. In the second case I think it's fine to build a query that goes directly against `Work` and turns the results into an `AcquisitionFeed`.

I'd also be okay with a method like `AcquisitionFeed.from_query(cls, query, facets, pagination)` which generalized this somewhat while allowing the caller to decide which query is best.

EdwinGuzman commented on api/admin/controller.py:
```diff
@@ -1557,9 +1568,12 @@ def _create_or_update_list(self, library, name, entries, collections, id=None):
-        new_urns = [entry.get("identifier_urn") for entry in entries]
+        new_urns = [entry.get("id") for entry in entries]
+        deleted_urns = [deletedEntry.get("id") for deletedEntry in deletedEntries]
         for entry in old_entries:
```
Updated.

EdwinGuzman commented:
Can you review this once more? Updated the usage of the `from_query` from core and updated tests to test that a feed is being returned from `custom_list`.

EdwinGuzman commented on api/admin/controller.py:
```diff
@@ -1600,26 +1608,36 @@ def custom_list(self, list_id):
+            feed = AcquisitionFeed(self._db, list.name, url, page_of_works, annotator)
+
+            # set_trace()
+            if len(list.entries) > 50 and pagination.has_next_page:
```
I moved this to a branch in core and it's no longer checking `len(list.entries)` https://github.com/NYPL-Simplified/server_core/pull/950/files#diff-eb69c9549013acc45b217e06734d0ec9R760

EdwinGuzman commented on api/opds.py:
```diff
@@ -755,18 +756,25 @@ def annotate_feed(self, feed, lane):
+                    type=OPDSFeed.ACQUISITION_FEED_TYPE,
+                    href=crawlable_url,
+                )
+                feed.add_link_to_feed(feed.feed, **crawlable_link)
```
The link this provides seems to return an empty OPDS feed with no entries, so maybe this is not really useful for the custom lists that are based on a WorkList?

leonardr replied:
It's not useful in this case, but adding it doesn't hurt anything, and it's useful for the other code that calls this method, so I would leave it in.

EdwinGuzman commented on tests/admin/test_controller.py:
```diff
@@ -2174,35 +2174,28 @@ def test_custom_lists_create(self):
+            eq_(2, len(feed.entries))
+            [crawlable_link] = [x['href'] for x in feed.feed['links']
+                              if x['rel'] == "http://opds-spec.org/crawlable"]
+            eq_(crawlable_link, "http://localhost/default/lists/%s/crawlable" % list.name)
```
You're not. I wasn't sure whether it was important to have or not. So I'll remove if it's causing confusion.

leonardr replied:
It's fine to have the link, but I would either remove it from the test or mention in a comment that this link isn't useful for this purpose.

EdwinGuzman commented on api/opds.py:
```diff
@@ -755,18 +756,25 @@ def annotate_feed(self, feed, lane):
+            if hasattr(lane, "customlists") and len(lane.customlists) == 1:
+                name = lane.customlists[0].name
+            elif (hasattr(lane, "customlist_ids") and lane.customlist_ids) and list:
+                name = list.name
```
Would something like this https://github.com/NYPL-Simplified/server_core/pull/950/commits/26db833cc5291f33287580196cb9049848787634 along with the latest update here, work?

EdwinGuzman commented on api/admin/controller.py:
```diff
@@ -1600,26 +1608,36 @@ def custom_list(self, list_id):
+                OPDSFeed.add_link_to_feed(feed=feed.feed, rel="previous", href=self.url_for("custom_list", after=pagination.previous_page.offset, library_short_name=library.short_name, list_id=list_id))
+
+
+            annotator.annotate_feed(feed, worklist)
```
Don't I need to have a worklist for `annotator = self.manager.annotator(worklist)`?

aslagle commented on api/admin/controller.py:
```diff
@@ -1504,11 +1506,36 @@ def custom_lists(self):
-            return self._create_or_update_list(library, name, entries, collections, id)
+            entries = self._getJSONFromRequest(flask.request.form.get("entries"))
+            collections = self._getJSONFromRequest(flask.request.form.get("collections"))
+            deletedEntries = self._getJSONFromRequest(flask.request.form.get("deletedEntries"))
```
I think this is always creating a new list so `deletedEntries` wouldn't apply, but if I'm missing something and it is needed, `deleteAll` should also be handled here.

EdwinGuzman replied:
No, that makes sense. If a new list is created then you'll only _really_ add new entries and the deleted ones we don't care about because those are just visually removed.

aslagle commented on api/admin/controller.py:
```diff
@@ -1560,6 +1560,9 @@ def _create_or_update_list(self, library, name, entries, collections, deletedEnt
         membership_change = False
 
+        if deleteAll:
+            list.entries = []
```
I think this might leave the CustomListEntry rows in the database, detached from their list. We may need to modify that relationship in the model so they get deleted. Otherwise this approach seems good.

aslagle replied:
Also, membership_change should be true in this case.

EdwinGuzman replied:
Can I just get every entry in the list and call `list.remove_entry` on it?

aslagle replied:
Yeah, I think that would also work.

EdwinGuzman replied:
Updated this.

aslagle commented on api/admin/controller.py:
```diff
@@ -1604,32 +1618,38 @@ def custom_list(self, list_id):
+            entries = self._getJSONFromRequest(flask.request.form.get("entries"))
+            collections = self._getJSONFromRequest(flask.request.form.get("collections"))
+            deletedEntries = self._getJSONFromRequest(flask.request.form.get("deletedEntries"))
+            deleteAll = flask.request.form.get("deletedEntries")
```
You're getting the wrong form parameter here.

EdwinGuzman commented:
@leonardr does this PR need any other changes before approval?

leonardr commented:
I went over it again and it looks good.

EdwinGuzman commented:
Thanks.
